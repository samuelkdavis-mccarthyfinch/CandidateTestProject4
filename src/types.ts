import { termSymbol } from './constants';

export interface Term {
  value: string;
  [termSymbol]: true;
}

export interface ConceptNode {
  '@id': string;
  '@type': NodeType.Node;
  label: string;
  language: Language;
  sense_label: string;
  term: string;
}

export interface Relation {
  '@id': RelationId;
  '@type': NodeType.Relation;
  label: string;
}

export interface Edge {
  '@id': string;
  '@type': NodeType.Edge;
  dataset: string;
  end: ConceptNode;
  license: string;
  rel: Relation;
  sources: Array<Object>;
  start: ConceptNode;
  surfaceText: null;
  weight: number;
}

export enum RelationId {
  IsA = '/r/IsA',
}

export enum Language {
  English = 'en',
}

export enum NodeType {
  Relation = 'Relation',
  Edge = 'Edge',
  Node = 'Node',
}

export interface TermRelationships {
  term: Term;
  ancestors: Record<string, boolean>;
}

export interface Relationships {
  [key: string]: Relationships;
}

export interface ConceptNetApi {
  getAncestralEdges: (term: Term) => Promise<Array<Edge>>;
  isParentOfATerm: (args: { parent: Term; child: Term }) => Promise<boolean>;
}
