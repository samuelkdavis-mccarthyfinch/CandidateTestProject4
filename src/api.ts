import { URL } from 'url';
import PromiseThrottle from 'promise-throttle';
import fetch from 'node-fetch';
import { MAX_REQUESTS_IN_A_MINUTE } from './constants';
import { RelationId, Edge, Term, ConceptNetApi } from './types';

interface Query {
  start?: string;
  end?: string;
  rel?: RelationId;
}

type ConceptNetResponse = {
  edges: Array<Edge>;
};

const getQueryUrl = (query: Query) => {
  const url = new URL('http://api.conceptnet.io/query');
  for (const [key, value] of Object.entries(query)) {
    url.searchParams.set(key, value);
  }

  return url.toString();
};

const urlGenerator = {
  termAncestors: (term: Term) => getQueryUrl({ start: term.value, rel: RelationId.IsA }),
  edgesBetweenChildAndParent: (args: { parent: Term; child: Term }) =>
    getQueryUrl({ start: args.child.value, end: args.parent.value, rel: RelationId.IsA }),
};

const promiseThrottle = new PromiseThrottle({
  requestsPerSecond: MAX_REQUESTS_IN_A_MINUTE / 60,
  promiseImplementation: Promise,
});

const getDataFromApi = async (url: string): Promise<ConceptNetResponse> => {
  console.log(`Making a request to ${url}`);
  const response = await promiseThrottle.add(() => fetch(url));
  const result = (await response.json()) as ConceptNetResponse;
  return result;
};

export const api: ConceptNetApi = {
  getAncestralEdges: async (term) => {
    const data = await getDataFromApi(urlGenerator.termAncestors(term));
    const edges = data.edges.filter((x) => x.rel['@id'] === RelationId.IsA);
    return edges;
  },
  isParentOfATerm: async (args) => {
    const data = await getDataFromApi(urlGenerator.edgesBetweenChildAndParent(args));
    return data.edges.length > 0;
  },
};
