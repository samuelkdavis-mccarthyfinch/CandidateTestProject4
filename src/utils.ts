import { termSymbol } from './constants';
import { Term, Language } from './types';

export const asTerm = (termString: string): Term => {
  return {
    value: termString,
    [termSymbol]: true,
  };
};

export const conceptToTerm = (concept: string, language: string = Language.English): Term => ({
  value: `/c/${language}/${concept.replace(/ /g, '_')}`,
  [termSymbol]: true,
});
