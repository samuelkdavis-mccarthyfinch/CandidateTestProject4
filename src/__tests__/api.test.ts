import { ConceptNetResponse, getApi } from '../api';
import fetch from 'node-fetch';
import { RelationId, Term } from '../types';
import { asTerm } from '../utils';

describe('api', () => {
  const mockFetch = jest.fn();
  const api = getApi({ fetch: mockFetch as unknown as typeof fetch });

  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe('getAncestralEdges', () => {
    test('correct result', async () => {
      const mockResponse = {
        edges: [
          {
            '@id': '1',
            rel: {
              '@id': RelationId.IsA,
            },
          },
          {
            '@id': '2',
            rel: {
              '@id': 'wrong',
            },
          },
          {
            '@id': '3',
            rel: {
              '@id': RelationId.IsA,
            },
          },
        ],
      };

      mockFetch.mockResolvedValueOnce({
        json: () => Promise.resolve(mockResponse),
      });

      const term = asTerm('mock_term');
      const result = await api.getAncestralEdges(term);
      expect(mockFetch.mock.calls).toMatchSnapshot();
      expect(result).toMatchSnapshot();
    });
  });

  describe('isParentOfATerm', () => {
    test('yes', async () => {
      const mockResponse = {
        edges: [
          {
            '@id': '1',
            rel: {
              '@id': RelationId.IsA,
            },
          },
        ],
      };

      mockFetch.mockResolvedValueOnce({
        json: () => Promise.resolve(mockResponse),
      });

      const term1 = asTerm('mock_term_1');
      const term2 = asTerm('mock_term_1');
      const result = await api.isParentOfATerm({ child: term1, parent: term2 });
      expect(mockFetch.mock.calls).toMatchSnapshot();
      expect(result).toBe(true);
    });

    test('no', async () => {
      const mockResponse = {
        edges: [],
      };

      mockFetch.mockResolvedValueOnce({
        json: () => Promise.resolve(mockResponse),
      });

      const term1 = asTerm('mock_term_1');
      const term2 = asTerm('mock_term_1');
      const result = await api.isParentOfATerm({ child: term1, parent: term2 });
      expect(mockFetch.mock.calls).toMatchSnapshot();
      expect(result).toBe(false);
    });
  });
});
