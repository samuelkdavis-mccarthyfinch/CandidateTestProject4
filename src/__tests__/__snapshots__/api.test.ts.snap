// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api getAncestralEdges correct result 1`] = `
Array [
  Array [
    "http://api.conceptnet.io/query?start=mock_term&rel=%2Fr%2FIsA",
  ],
]
`;

exports[`api getAncestralEdges correct result 2`] = `
Array [
  Object {
    "@id": "1",
    "rel": Object {
      "@id": "/r/IsA",
    },
  },
  Object {
    "@id": "3",
    "rel": Object {
      "@id": "/r/IsA",
    },
  },
]
`;

exports[`api isParentOfATerm no 1`] = `
Array [
  Array [
    "http://api.conceptnet.io/query?start=mock_term_1&end=mock_term_1&rel=%2Fr%2FIsA",
  ],
]
`;

exports[`api isParentOfATerm yes 1`] = `
Array [
  Array [
    "http://api.conceptnet.io/query?start=mock_term_1&end=mock_term_1&rel=%2Fr%2FIsA",
  ],
]
`;
